Use wheelDelta instead of delta in the scroll handler

Ben3eeE opened the pull request:
In Atom 1.19 we changed the scroll handler to use deltaX and deltaY instead of wheelDeltaX/wheelDeltaY. wheelDelta is larger so this caused the scrolling speed to slow down. This change in speed was especially noticeable on Linux.

Fixes https://github.com/atom/atom/issues/15567

# Test plan

* Scrolling on macOS with a mouse
  * [x]  Scroll in every direction
  * [x]  Scroll in every direction while holding shift
* Scrolling on macOS with a touchpad
  * [x]  Scroll in every direction
  * [x]  Scroll in every direction while holding shift
  * [x]  Slightly scroll down/up, ensuring that for small values of `wheelDeltaY * scrollSensitivity` we still honor the scrolling.
  * [x]  Slightly scroll left/right, ensuring that for small values of `wheelDeltaX * scrollSensitivity` we still honor the scrolling.
* Scrolling on Windows with a mouse
  * [x]  Scroll in every direction
  * [x]  Scroll in every direction while holding shift
* Scrolling on Windows with a touchpad
  * [ ]  Scroll in every direction
  * [ ]  Scroll in every direction while holding shift
  * [ ]  Slightly scroll down/up, ensuring that for small values of `wheelDeltaY * scrollSensitivity` we still honor the scrolling.
  * [ ]  Slightly scroll left/right, ensuring that for small values of `wheelDeltaX * scrollSensitivity` we still honor the scrolling.
* Scrolling on Linux with a mouse
  * [X] Scroll in every direction
  * [X]  Scroll in every direction while holding shift
* Scrolling on Linux with a touchpad
  * [x] Scroll in every direction
  * [x]  Scroll in every direction while holding shift
  * [x]  Slightly scroll down/up, ensuring that for small values of `wheelDeltaY * scrollSensitivity` we still honor the scrolling.
  * [x]  Slightly scroll left/right, ensuring that for small values of `wheelDeltaX * scrollSensitivity` we still honor the scrolling.


/cc: @nathansobo @as-cii

Ben3eeE committed:
Use wheelDelta instead of delta

In Atom 1.19 we changed the scroll handler to use deltaX and deltaY
instead of wheelDeltaX/wheelDeltaY. wheelDelta is larger so this caused
the scrolling speed to slow down. This change in speed was especially
noticable on Linux

50Wliu commented:
Furthermore, why are we using the [deprecated, non-standard `mousewheel` event](https://developer.mozilla.org/en-US/docs/Web/Events/mousewheel)?
 https://github.com/atom/atom/blob/42657780d51c357d693328393efebf922ccd7207/src/text-editor-component.js#L481

The standard event is [`wheel`](https://developer.mozilla.org/en-US/docs/Web/Events/wheel), but that event does not have `wheelDeltaX` or `wheelDeltaY`.  It might, however, report more accurate values for `deltaX` and `deltaY`.

50Wliu commented:
See also MDN's documentation on what the delta values actually mean.  https://developer.mozilla.org/en-US/docs/Web/Events/wheel#The_delta_values

Ben3eeE commented:
> Furthermore, why are we using the deprecated, non-standard mousewheel event?

I have also noticed this and I am not sure why we use `mousewheel`. But the two events `wheel` and `mousewheel` currently seem to be identical. I couldn't find any difference between the two.

> See also MDN's documentation on what the delta values actually mean. https://developer.mozilla.org/en-US/docs/Web/Events/wheel#The_delta_values

This is what firefox does but chrome doesn't follow this at all.

50Wliu commented:
> This is what firefox does but chrome doesn't follow this at all.

Well, that sucks :/.  Is that for `mousewheel` or `wheel` as well?

Ben3eeE commented:
>  Is that for mousewheel or wheel as well?

Since mousewheel and wheel are identical it's for both. We could probably use  deltaX/deltaY and the standard event but currently scrolling is not really usable on Linux so I propose we change it back to the way it was in Atom 1.18 that worked and then figure it out later.

Searching around I found a bunch of issues on the chromium project about delta values being too small on Linux compared to Windows and all were closed as wontfix. We didn't get issues about scrolling on Linux prior to Atom 1.19 as we do now so I think it's fair to go back to the previous behavior for now.

I know that the tests are currently failing because I didn't change them at all.

Ben3eeE committed:
Make the tests pass

Ben3eeE committed:
And don't focus the test

as-cii commented:
Looks good to me if it fixes the scrolling issues on Linux. Have you tested all scenarios revolving around scrolling? e.g.

* Scrolling on macOS
  * Scroll in every direction
  * Scroll in every direction while holding shift
  * Slightly scroll down/up, ensuring that for small values of `wheelDeltaY * scrollSensitivity` we still honor the scrolling.
  * Slightly scroll left/right, ensuring that for small values of `wheelDeltaX * scrollSensitivity` we still honor the scrolling.
* Scrolling on Windows
  * Scroll in every direction
  * Scroll in every direction while holding shift
  * Slightly scroll down/up, ensuring that for small values of `wheelDeltaY * scrollSensitivity` we still honor the scrolling.
  * Slightly scroll left/right, ensuring that for small values of `wheelDeltaX * scrollSensitivity` we still honor the scrolling.
* Scrolling on Linux
  * Scroll in every direction
  * Scroll in every direction while holding shift
  * Slightly scroll down/up, ensuring that for small values of `wheelDeltaY * scrollSensitivity` we still honor the scrolling.
  * Slightly scroll left/right, ensuring that for small values of `wheelDeltaX * scrollSensitivity` we still honor the scrolling.

I am also a little unclear still on the difference between `mousewheel` vs. `wheel`, and between `wheelDelta` vs `delta`. @50Wliu @Ben3eeE: can you help me understand how and why they differ?

Nice work as usual, @Ben3eeE. ⚡️

Ben3eeE commented:
Thanks for the feedback and test plan @as-cii :100: :sparkles: 

I would like to change the test plan so we test scrolling with a touchpad **and** mouse. I've updated your test plan and added it to the body of the PR. I fear that while this improves the experience with scrolling with a mouse on Linux it might make touchpads worse. I hope the current test plan will ensure we don't regress either touchpads or mice and will tell us if we should implement a different solution.

---

> I am also a little unclear still on the difference between mousewheel vs. wheel, and between wheelDelta vs delta. @50Wliu @Ben3eeE: can you help me understand how and why they differ?

In Chrome there is no difference between `mousewheel` and `wheel`. `mousewheel` is a non standard deprecated event and `wheel` is the standard event. But all properties of the events are the same. I think `mousewheel` is just kept around for compatibility and that `wheelDelta` was added to the `wheel` event for the same reason. But this is just me guessing.

The difference between `wheelDelta` and `delta` is more complicated. `wheelDelta` is a non standard property that was around before the DOM3 standard event.

**According to MDN:**

**`delta`:**
Reports delta in one of the available [`deltaModes`](https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent/deltaMode). To my knowledge Chromium never reports delta using `DOM_DELTA_LINE` but it can report using `DOM_DELTA_PAGE`. If we don't merge this PR and instead implement scrolling using `delta` we should make sure we respect all available modes in the scroll handler.
The problem with this delta is that it is really small on Linux so it doesn't seem like it works well on all platforms. On Windows the difference using the default OS settings is 20 less down from 120 to 100. On Linux the delta was 53 when I tested using a standard Ubuntu 16.04 with gnome. This difference makes scrolling very slow on Linux and the sudden change in speed resulted in issues from users.

**`wheelDelta`:**
* On Windows this delta is based on the delta value of `WM_MOUSEWHEEL`
* On Linux it is always `+/-120` making it "The same as IE and Chrome on Windows"
* On macOS it's complicated and depends on the device used and what scrolling modes it supports

Because of the difference between `delta` and `wheelDelta` I feel that testing with both a mouse and a touchpad is required. I fear that this might make scrolling with a touchpad worse on Linux because the change is always 120?

---

I am not sure how I feel about this solution and I think we could maybe further improve the experience and better respect the OS settings by using the standard `delta` property. Regardless of the solution we pick we should go through the same test plan to ensure that scrolling is fast and feels good on all platforms with both a mouse and a touchpad.

nathansobo commented:
It could be that using `delta` and interpreting the `deltaMode` would lead to a more correct solution, but if people were happy with things when we were using the old deprecated property I think we should return to using it. There may be more for us to learn here. So I say we do what we need to do to relieve the immediate pain and optionally think harder about using the most standards-compliant solution possible in the future.

Ben3eeE commented:
@nathansobo Thank you for the feedback ✨ 

I am going to need some help going through this test plan because I don't have the required hardware. I can test Windows and Linux using a mouse but I don't have access to a touchpad or macOS.

We can post in https://github.com/atom/atom/issues/11976 about switching to `delta` to improve the native experience and share our current knowledge about these properties. Maybe someone from the community wants to pick it up.

nathansobo commented:
Once you confirm on Linux and Windows I can test on macOS.

Ben3eeE commented:
@nathansobo I have tested Windows and Linux with a mouse and can confirm that scrolling works.

I can also confirm that this goes back to the old behavior of ignoring the scroll sensitivity settings that you can configure in Windows. And that we don't slow down scrolling to one pixel per notch when configured to page scrolling because we don't honor `deltaMode` in the old handler.

nathansobo commented:
> I can also confirm that this goes back to the old behavior of ignoring the scroll sensitivity settings that you can configure in Windows. And that we don't slow down scrolling to one pixel per notch when configured to page scrolling because we don't honor deltaMode in the old handler.

Not sure if I'm interpreting your meaning clearly, but this sounds like a bad thing?

Ben3eeE commented:
> Not sure if I'm interpreting your meaning clearly, but this sounds like a bad thing?

I don't know. It is an expected thing because `delta` changes with the OS settings and `wheelDelta` does not. And as I have said if we use `delta` we can do a better job at respecting the OS settings (At least on Windows with a mouse, I have no idea about other platforms/hardware). But our current implementation doesn't respect the OS setting on Windows because `delta` based scrolling was added by mistake.

So while the speed would increase if you increased it in the Windows settings it would increase by an amount that does not correspond to the new setting.

nathansobo commented:
Okay. I'm a bit confused, but I think we should improve the experience on Linux if it doesn't hurt the experience on Windows and take a deeper look at scrolling. Sounds like we need to interpret the `deltaMode` parameter.

nathansobo commented:
@rsese Do you think you could run through the test plan stated in the body of this PR some time soon? I'd love to get this merged soon and relieve some pain for Linux users in #15567.

rsese commented:
Yep I can give this a try today @nathansobo :+1:   I don't have a solid working build of Atom on Windows though so probably won't get to that last Windows + touchpad section (and there's no artifact to download from AppVeyor it looks like?).

One question - can you clarify the behavior I should see for these items?

>   - [ ] Slightly scroll down/up, ensuring that for small values of wheelDeltaY * scrollSensitivity we still honor the scrolling
>   - [ ] Slightly scroll left/right, ensuring that for small values of wheelDeltaX * scrollSensitivity we still honor the scrolling.

Is this just verifying that very slight scrolling results in scrolling at all?

nathansobo commented:
> Is this just verifying that very slight scrolling results in scrolling at all?

I think so. And that it doesn't scroll by too much maybe? Linus would have the best answer but I think he's out for a bit.

nathansobo commented:
AppVeyor failure looks like a flake so I kicked it.

rsese commented:
> I think so. And that it doesn't scroll by too much maybe? Linus would have the best answer but I think he's out for a bit.

:+1:

From this branch, I ran through the macOS section (touchpad + mouse) and the Linux + touchpad section (Ubuntu VM).  To compare, I ran through things from Atom `1.24.0-dev-4f6e8ed5f` first.

On macOS, scrolling in every direction with a mouse works fine and it does feel a tiny bit faster and additionally, you scroll ~2 times as much now.  With the touchpad I saw the same thing - feels a bit faster and you scroll ~2 times as much.  If anyone changed their scroll sensitivity to 80 because of https://github.com/atom/atom/issues/15567, they'll end up scrolling quite a bit and may switch back to the default 40.  Unless I mis-understood, touchpad scrolling while holding shift doesn't change the scrolling behavior (like it does with the mouse to scroll horizontally).  Scrolling extremely slowly did scroll with the touchpad and mouse.

On Linux + touchpad, scrolling in every direction worked fine (had to hold shift to scroll horizontally with the touchpad on Ubuntu).  Scrolling did feel faster and like with macOS, you scroll about ~2 as much.  You don't scroll as much as you do on macOS, but definitely a better experience when running off of master which is pretty sluggish.  Scrolling extremely does scroll eventually though Atom doesn't scroll continuously while I scroll on the touchpad like it does with macOS.

rsese commented:
Checked off the macOS and the Linux + touchpad section - looks like AppVeyor failed again, I'll see if I can get my Windows VM building later tonight.

rsese commented:
Just saw AppVeyor passed but I ended up getting Atom building on my Windows vm - went through the Windows + touchpad section and didn't notice a difference between master and this branch.  Not sure if that was expected with the changes in this PR?

nathansobo commented:
I think it's good that the behavior is the same. It means we just make Linux better.
